dts: bindings: dma: gd32: Redesign DMA node's cell properties

soburi opened the pull request:
From #47504 review, split out to this PR about gd32-dma related changes.

From result of discussion in this PR, the cell property specs following STM32's interface specs is preferable.

This PR make following changes...

- Add `config` cell property to gd,gd32-dma
- Add mem2mem property to indicate it supports memory to memory transfer
- Split `gd,gd32-dma-v1` from `gd,gd32-dma` for supporting F4XX specific features (FIFO is defined, but not implemented)
- Use dma_slot for peripheral selectiong instead of linked_channel.
- Add/Correct overlay files for dma/loop_transfer test.

I verify the dma/loop_transfer test passed with gd32f407v_start and longan_nano.
Other boards case, I only checked it can compile.

soburi committed:
drivers: dma: dma_gd32: Use dma_slot for peripheral request

Use dma_slot for peripheral request instead of linked_channel.
This is a more suitable usage as described in dma_config.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

galak commented on dts/bindings/dma/gd,gd32-dma.yaml:
```diff
@@ -4,6 +4,32 @@
+
+  channel: Select channel for data transmitting
+  priority: Set priority for this channel
+  peripheral: Select peripheral to connect DMA (Only for GD32F4xx series)
```
Can `peripheral` not be determined by the device node that has the `dmas` property?

soburi replied:
Exactly as you pointed out, it can determine which peripheral to use in the client code that use the DMA.
And it doesn't have to be configurable.
I removed it.

soburi replied:
I think it can't determine.

The peripheral assignment is tied to the DMA channel.
For example, from  [GD32f450 manual](https://www.gigadevice.com/manual/gd32f450xxxx-user-manual/) Page 211 Table 10-6,
SPI0-RX can use with pair of channel-0 and peripheral-3 or channel-2 and peripheral-3.
Channel is already configurable with the property.
Peripheral is also needs to be configurable with property to follow channel configuration.

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/pull/48756#issuecomment-1209025306

I have changed design since this discussion.
So I assume this Change Request is no longer needed. Could you check it please?

galak reviewed (dismissed):
Question on devicetree binding.

gmarull commented:
I think GD32 dma bindings must align with STM32, see

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/st,stm32-dma.yaml

and Zephyr counterpart.

soburi commented:
Hi @gmarull , @galak 

Sorry for late response.

> I think GD32 dma bindings must align with STM32, see
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/st,stm32-dma.yaml
> 
> and Zephyr counterpart.

I reworked this PR to make it compatible with STM32's dts spec.

I think the STM32's channel-config cell design is a bit complex.
(And almost of these should determine by what peripheral using.)
But making it compatible with STM32 is a preferable policy, and I followed it.

Could you tell me your opinions?

carlescufi commented:
@galak @gmarull ping

gmarull commented on include/zephyr/drivers/dma/dma_gd32.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef ZEPHYR_INCLUDE_DRIVERS_DMA_GD32_H_
+#define ZEPHYR_INCLUDE_DRIVERS_DMA_GD32_H_
+
+#define GD32_DMA_CONFIG_DIRECTION(config)	     ((config >> 6) & 0x3)
+#define GD32_DMA_CONFIG_PERIPH_ADDR_INC(config)	     ((config >> 9) & 0x1)
+#define GD32_DMA_CONFIG_MEMORY_ADDR_INC(config)	     ((config >> 10) & 0x1)
+#define GD32_DMA_CONFIG_PERIPH_WIDTH(config)	     ((config >> 11) & 0x3)
+#define GD32_DMA_CONFIG_MEMORY_WIDTH(config)	     ((config >> 13) & 0x3)
+#define GD32_DMA_CONFIG_PERIPHERAL_INC_FIXED(config) ((config >> 15) & 0x1)
+#define GD32_DMA_CONFIG_PRIORITY(config)	     ((config >> 16) & 0x3)
+
+#define GD32_DMA_FEATURES_FIFO_THRESHOLD(features) (features & 0x3)
```
is this needed in a public header?

soburi replied:
Sorry for response delayed.

I think it is needed.
Others than DMA driver uses this header.
The cell properties are handled by a driver that uses the DMA.

gmarull reviewed (commented):
just one question

soburi commented:
Hi, @gmarull @galak @carlescufi 

Could you re-check this PR if you can take a while.

Point to discussion:
IMO, The GD32 DMA property cell's specs should be identical or similar to the STM32.
From some conversations with @galak, I also understanding that simple specs are preferable.
On the other hand, The STM32's DMA property cell specs are a little complex.
(Because the specs make it able to set all properties. So we need to provide a header file that define the property bits.)
But in this case, I think it is better to make the specs compatible with STM32's one for compatibility reasons.


If I can, I want to proceed with this PR and #47504 to bring it into v3.3.

gmarull commented on dts/bindings/dma/gd,gd32-dma.yaml:
```diff
@@ -4,6 +4,79 @@
+        pinctrl-names = "default";
+        cs-gpios = <&gpioa 4 GPIO_ACTIVE_LOW>;
+
+        dmas = <&dma1 0 3 0 0>, <&dma1 5 3 (GD32_DMA_PRIORITY_HIGH) 0>
```
nit here: `GD32_DMA_PRIORITY_HIGH` should not be in parenthesis

soburi replied:
Thank you for pointing out.
Fixed it.

gmarull commented on dts/bindings/dma/gd,gd32-dma.yaml:
```diff
@@ -4,6 +4,79 @@
+  Each is named "rx" and "tx".
+  The first cell assign channel 3 to receive, channel 5 assigned to transmit.
+  The second cell is slot. Both channels select 3.
+  Slot is effective only for GD32F4XX series. ignored in otherwise.
+  What the slot number '3' meaning depends on the DMA controller and channel.
+  See Hardware manual.
```
Note: it seems that we're in front of different DMA IPs if that is true, have you considered using different compatibles (as ST does)? Not sure if it is worth here, but if F4XX series is the only one requiring `slot`, it could have a special binding with slot cell.

soburi replied:
@gmarull , @cameled 

I split the configuration that is for F4xx.

- gd,gd32-dma  (channel, channel-config)
This is a basic configuration.
The stm32 implementation, this configuration is named gd,gd32-dma-v2bis.
It is a confusable name.
I'd like to keep it simple.
But it might be better to match the stm32 naming.
I can change the name quickly if so on.

I would like to hear your opinion on this point. > @gmarull @cameled @nandojve 


- gd,gd32-dma-v1 [channel, slot, channel-config, feature]
This is configuration for F4XX series.
This is following stm,stm32-dma-v1 cells specs.

For reference, stm32-dma has three configurations.

- stm,stm32-dma-v1 [channel, slot, channel-config, feature]
- stm,stm32-dma-v2 [channel, slot, channel-config]
- stm,stm32-dma-v2bis [channel, channel-config]

(slot is in an awkward position...)

gmarull reviewed (commented):
thanks for the proposal, I have a couple of minor comments.

soburi committed:
dts: bindings: gd32-dma: add config cell property

Add config cell property to gd,gd32-dma.
For supporting hardware variation, Splitting base definition
to gd,gd32-dma-base.yaml.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: bindings: dma: gd32: split gd,gd32-dma-v1 for support F4xx feature

Split gd,gd32-dma-v1 from gd,gd32-dma to support F4xx specific features.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: dma: loop_transfer: Correct DMA definition of gd32f350r_eval

The overlay file overrides not existing node `dma0`.
Correct it to `dma`.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: dma: loop_transfer: add overlay for GD32 boards

Add overlay file for gd32_e507z_eval, gd32f407v_start,
gd32f450v_start, gd32f450v_start and longan_nano_lite.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: bindings: gd32-dma-base: add `gd,mem2mem` property

Add `gd,mem2mem` property to indicate the DMA controller supports
memory to memory transfer.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: gd32: add `dma` as supported feature

Add `dma` as supported feature to yaml configuration.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented on dts/bindings/dma/gd,gd32-dma-v1.yaml:
```diff
@@ -0,0 +1,91 @@
+dma-cells:
+  - channel
+  - slot
+  - channel-config
```
It comes from stm32.
Please see also https://github.com/zephyrproject-rtos/zephyr/pull/48756?notification_referrer_id=NT_kwDOAAb-KbE0MTQ0NjQ5MDUxOjQ1ODI4MQ#discussion_r1045202061

nandojve replied:
I would like suggest that maybe **channel-config** could be renamed to **config** or even **settings* and *features* renamed to *threshold* or even *fifo-threshold*, since it only configure that. In the case of features, I did not understand what is the purpose of it. I mean, what for is that threshold? You need to explain that.

soburi replied:
hi, @nandojve 

Thank you for proposal.
I applied it. Rename the cell names as
channel-config -> config
feature  -> fifo-threshold

nandojve commented on dts/bindings/dma/gd,gd32-dma-v1.yaml:
```diff
@@ -0,0 +1,91 @@
+                 - 0x2: high
+                 - 0x3: very high
+
+  fifo-threshold: A 32bit bitfield value specifying FIFO threshold
+    - bit 0-1:   DMA FIFO threshold selection
```
What is the purpose of this cell? Is this a trigger? What FIFO is this? Is it part of peripheral side, DMA side? At devicetre tree bindings you must state the meaning in a way that maintainers don't need to look at datasheet. The obvious cells like channel and slot don't require to much attention but platform specific we need good explanation.

soburi replied:
Thank you for checking.
I update the description about the `fifo-threshold`.
This cell is a configuration for the burst transfer feature (but not yet supported), 
clarified how it works with other parameters.

gmarull commented on dts/arm/gigadevice/gd32e50x/gd32e50x.dtsi:
```diff
@@ -411,6 +411,7 @@
 				     <15 0>, <16 0>, <17 0>;
 			clocks = <&cctl GD32_CLOCK_DMA0>;
 			dma-channels = <7>;
+			mem2mem;
```
nit: use vendor prefix for this prop `gd,mem2mem` (we likely need to fix a few others already in-tree)

soburi replied:
Thank you for pointing out.
I Fixed it.
